src/core_deku/contract_ffi.ml

```diff
@@ -1,55 +1,172 @@
+      store (call data)
+    | OWN_TICKET ->
+      let _, ticket_handle = R.read_ticket memory writer_addr int32_offset in
+      let store handle1 = W.write_ticket ~writer_addr handle1 memory in
+      store (call ticket_handle)
+    | JOIN_TICKETS ->
+      let addr, ticket_handle = R.read_ticket memory writer_addr int32_offset in
+      let _, ticket_handle2 = R.read_ticket memory addr int32_offset in
+      let store handle1 = W.write_ticket ~writer_addr handle1 memory in
+      store (call (ticket_handle, ticket_handle2))
+    | TRANSACTION ->
+      let addr =
+        Bytes.get_int32_le
+          (Memory.load_bytes memory
+             ~address:(Int64.add writer_addr (int32_offset |> Int64.of_int))
+             ~size:4)
+          0
+        |> Int64.of_int32 in
+      let arg =
+        if addr = -1L then
+          Bytes.empty
+        else
+          let size =
+            Bytes.get_int64_le
+              (Memory.load_bytes memory ~address:addr ~size:8)
+              0
+            |> Int64.to_int in
+          let bytes =
+            Memory.load_bytes memory ~address:(Int64.add addr 8L) ~size in
+          bytes in
+      let addr, ticket_handle =
+        R.read_ticket memory writer_addr (int32_offset * 2) in
+      let addr, amount = R.read_amount memory addr int32_offset in
+      let _, destination = R.read_address memory addr int64_offset in
+      let store int =
+        let buf = Bytes.make 4 '\x00' in
+        Bytes.set_int32_le buf 0 (Int32.of_int int);
+        Memory.store_bytes memory ~address:writer_addr ~content:buf in
+      let called = call (arg, (ticket_handle, amount), destination) in
+      store called
+    | GET_CONTRACT_OPT ->
+      let _, address = R.read_address memory writer_addr int32_offset in
+      let store addr = W.write_some ~writer_addr addr W.write_address memory in
+      store (call address)
 
-   let syscall (type a) mem (module S : CTX) =
-     let module T = S in
-     fun x ->
-       match Array.get mem x with
-       | 0 -> load (FAILWITH 0) |> T.failwith
-       | 1 -> load (SENDER 1) |> T.sender
-       | _ -> failwith "test"
-
-   module S : CTX = struct
-     let sender = Address.of_string "test" |> Option.get
-     let failwith _ = ()
-     let sender f = f sender
-   end
-   let res = syscall [|0|] (module S) 1 *)
+let syscall (module S : Contract_context.CTX) mem x =
+  let open Wasm_vm in
+  match
+    Bytes.get_int32_le (Memory.load_bytes mem ~address:x ~size:4) 0
+    |> Int32.to_int
+  with
+  | 0 -> load mem x S.self SELF
+  | 1 -> load mem x S.source SOURCE
+  | 2 -> load mem x S.sender SENDER
+  | 3 -> load mem x S.read_ticket READ_TICKET
+  | 4 -> load mem x S.split_ticket SPLIT_TICKET
+  | 5 -> load mem x S.own_ticket OWN_TICKET
+  | 6 -> load mem x S.join_tickets JOIN_TICKETS
+  | 7 -> load mem x S.get_contract_opt GET_CONTRACT_OPT
+  | 8 -> load mem x S.transaction TRANSACTION
+  | _ -> failwith "unimplemented"
```

Comment:
failing would make possible for an external agent to crash the node by running a unimplemented syscall.

Reply:
We catch all errors inside runtime.ml of wasm_vm library

Reply:
So that should be a defined exception and not a failure

Reply:
what's the point of defining an exception if the only thing we can do about it is to abort execution?